File: Semester3/MAP/laboratoare/lab2/TaskManager/src/Decorator/StrategyTaskRunner.java
package Decorator;

import Domain.Container;
import Domain.Task;
import Factory.Strategy;
import Factory.TaskContainerFactory;

import java.util.EmptyStackException;

public class StrategyTaskRunner implements TaskRunner {
    private Container container;

    public StrategyTaskRunner(Strategy strategy) {
        container = TaskContainerFactory.getInstance().createContainer(strategy);
    }

    @Override
    public void executeOneTask() {
        if(!container.isEmpty())
            container.remove().execute();
        else
            throw new EmptyStackException();
    }

    @Override
    public void executeAll() {
        while(hasTask()){
            executeOneTask();
        }
    }

    @Override
    public void addTask(Task task) {
        container.add(task);
    }

    @Override
    public boolean hasTask() {
        return !container.isEmpty();
    }
}

File: Semester3/MAP/Seminare/Sem10/Sem10/Task/StrategyTaskRunner.cs
namespace Sem10;

public class StrategyTaskRunner : TaskRunner
{
    private Container container;

    public StrategyTaskRunner(Strategy strategy)
    {
        container = TaskContainerFactory.getInstance().createContainer(strategy);
    }

    public void executeOneTask()
    {
        if (!container.isEmpty())
        {
            Task task = container.remove();
            task.execute();
        }
    }

    public void executeAll()
    {
        while(!container.isEmpty())
            executeOneTask();
    }

    public void addTask(Task task)
    {
        container.add(task);
    }

    public bool hasTasks()
    {
        return !container.isEmpty();
    }
    
}